File: BluffinServer/src/serverLobby/ServerClientLobby.java
package serverLobby;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.Socket;
import java.net.SocketException;

import protocolLobby.LobbyConnectCommand;
import protocolLobby.LobbyCreateTableCommand;
import protocolLobby.LobbyDisconnectCommand;
import protocolLobby.LobbyListTableCommand;
import protocolLobbyTools.LobbyServerSideAdapter;
import protocolLobbyTools.LobbyServerSideObserver;
import protocolTools.IPokerCommand;

/**
 * This class represents a client for ServerLobby.
 */
public class ServerClientLobby extends Thread
{
    private String m_playerName = "?";
    private final ServerLobby m_lobby;
    private final LobbyServerSideObserver m_commandObserver = new LobbyServerSideObserver();
    
    // Communications with the client
    private Socket m_socket = null;
    private PrintWriter m_toClient;
    private BufferedReader m_fromClient;
    
    public ServerClientLobby(Socket socket, ServerLobby lobby)
    {
        m_socket = socket;
        m_lobby = lobby;
        try
        {
            m_toClient = new PrintWriter(socket.getOutputStream(), true /* autoFlush */);
            m_fromClient = new BufferedReader(new InputStreamReader(socket.getInputStream()));
        }
        catch (final IOException e)
        {
            e.printStackTrace();
        }
    }
    
    public boolean isConnected()
    {
        return m_socket.isConnected() && !m_socket.isClosed();
    }
    
    /**
     * Send message to client.
     * 
     * @param p_msg
     *            - Message to send.
     */
    protected void sendMessage(String p_msg)
    {
        System.out.println("Server SEND to " + m_playerName + " [" + p_msg + "]");
        m_toClient.println(p_msg);
    }
    
    protected String receive() throws IOException
    {
        final String line = m_fromClient.readLine();
        m_commandObserver.receiveSomething(line);
        return line;
    }
    
    protected void send(IPokerCommand p_msg)
    {
        sendMessage(p_msg.encodeCommand());
    }
    
    @Override
    public void run()
    {
        initializeCommandObserver();
        while (isConnected())
        {
            try
            {
                receive();
            }
            catch (final SocketException e)
            {
                System.out.println("Connection lost with " + m_playerName);
                return;
            }
            catch (final Exception e)
            {
                e.printStackTrace();
            }
            
        }
    }
    
    private void initializeCommandObserver()
    {
        m_commandObserver.subscribe(new LobbyServerSideAdapter()
        {
            @Override
            public void commandReceived(String command)
            {
                System.out.println("Server RECV from " + m_playerName + " [" + command + "]");
            }
            
            @Override
            public void connectCommandReceived(LobbyConnectCommand command)
            {
                m_playerName = command.getPlayerName();
                final boolean ok = !m_lobby.isNameUsed(m_playerName);
                sendMessage(command.encodeResponse(ok));
                if (ok)
                {
                    m_lobby.addName(m_playerName);
                }
            }
            
            @Override
            public void disconnectCommandReceived(LobbyDisconnectCommand command)
            {
                m_lobby.removeName(m_playerName);
                try
                {
                    m_socket.close();
                }
                catch (final IOException e)
                {
                    e.printStackTrace();
                }
            }
            
            @Override
            public void createTableCommandReceived(LobbyCreateTableCommand command)
            {
                sendMessage(command.encodeResponse(m_lobby.createTable(command)));
            }
            
            @Override
            public void listTableCommandReceived(LobbyListTableCommand command)
            {
                sendMessage(command.encodeResponse(m_lobby.listTables()));
            }
        });
        
    }
}

File: C#Integration/BluffinPokerServer/ServerClientLobby.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using EricUtility.Networking.Commands;
using PokerProtocol.Observer;
using System.Net.Sockets;
using PokerProtocol.Commands.Lobby;

namespace BluffinPokerServer
{
    public class ServerClientLobby : CommandTCPCommunicator<LobbyServerCommandObserver>
    {
        private string m_PlayerName = "?";
        private readonly ServerLobby m_Lobby;

        public ServerClientLobby(TcpClient client, ServerLobby lobby)
            : base(client)
        {
            m_Lobby = lobby;
        }
        protected override void InitializeCommandObserver()
        {
            m_CommandObserver.CommandReceived += new EventHandler<StringEventArgs>(m_CommandObserver_CommandReceived);
            m_CommandObserver.IdentifyCommandReceived += new EventHandler<CommandEventArgs<IdentifyCommand>>(m_CommandObserver_IdentifyCommandReceived);
            m_CommandObserver.DisconnectCommandReceived += new EventHandler<CommandEventArgs<DisconnectCommand>>(m_CommandObserver_DisconnectCommandReceived);
            m_CommandObserver.CreateTableCommandReceived += new EventHandler<CommandEventArgs<CreateTableCommand>>(m_CommandObserver_CreateTableCommandReceived);
            m_CommandObserver.ListTableCommandReceived += new EventHandler<CommandEventArgs<ListTableCommand>>(m_CommandObserver_ListTableCommandReceived);
        }
        protected override void Send(string line)
        {
            Console.WriteLine("Server SEND to " + m_PlayerName + " [" + line + "]");
            base.Send(line);
        }

        void m_CommandObserver_CommandReceived(object sender, StringEventArgs e)
        {
            Console.WriteLine("Server RECV from " + m_PlayerName + " [" + e.Str + "]");
        }

        void m_CommandObserver_IdentifyCommandReceived(object sender, CommandEventArgs<IdentifyCommand> e)
        {
            IdentifyCommand c = e.Command;
            m_PlayerName = c.Name;
            bool ok = !m_Lobby.NameUsed(m_PlayerName);
            Send(c.EncodeResponse(ok));
            if (ok)
                m_Lobby.AddName(m_PlayerName);
        }

        void m_CommandObserver_DisconnectCommandReceived(object sender, CommandEventArgs<DisconnectCommand> e)
        {
            DisconnectCommand c = e.Command;
            m_Lobby.RemoveName(m_PlayerName);
            Close();
        }

        void m_CommandObserver_CreateTableCommandReceived(object sender, CommandEventArgs<CreateTableCommand> e)
        {
            CreateTableCommand c = e.Command;
            Send(c.EncodeResponse(m_Lobby.CreateTable(c)));
        }

        void m_CommandObserver_ListTableCommandReceived(object sender, CommandEventArgs<ListTableCommand> e)
        {
            ListTableCommand c = e.Command;
            Send(c.EncodeResponse(m_Lobby.ListTables()));
        }
    }
}
